File: apps/macos/Lumina/Features/Reader/SegmentBoundaryOffset.swift
import Foundation

/// Maps NSTextView UTF-16 indices to Python `len(str)` / Unicode scalar offsets
/// used by `POST .../boundary` `left_char_count`.
enum SegmentBoundaryOffset {
    static func unicodeOffset(utf16Index: Int, in text: String) -> Int {
        let ns = text as NSString
        guard ns.length > 0 else { return 0 }
        var index = min(max(0, utf16Index), ns.length)
        if index > 0, index < ns.length {
            let unit = ns.character(at: index)
            if UTF16.isTrailSurrogate(unit) {
                index -= 1
            }
        }
        return (ns.substring(to: index) as String).unicodeScalars.count
    }

    static func utf16Index(forUnicodeOffset offset: Int, in text: String) -> Int {
        if offset <= 0 { return 0 }
        var remaining = offset
        var utf16 = 0
        for scalar in text.unicodeScalars {
            if remaining <= 0 { break }
            utf16 += String(scalar).utf16.count
            remaining -= 1
        }
        return utf16
    }

    static func split(_ text: String, unicodeOffset: Int) -> (String, String) {
        let total = text.unicodeScalars.count
        let clamped = min(max(0, unicodeOffset), total)
        let left = String(String.UnicodeScalarView(text.unicodeScalars.prefix(clamped)))
        let right = String(String.UnicodeScalarView(text.unicodeScalars.dropFirst(clamped)))
        return (left, right)
    }
}

File: apps/windows/Lumina/Features/Reader/SegmentBoundaryOffset.cs
using System.Collections.Generic;
using System.Linq;

namespace Lumina.Features.Reader;

/// <summary>
/// Maps UTF-16 caret indices to Python <c>len(str)</c> / Unicode scalar offsets
/// used by <c>POST .../boundary</c> <c>left_char_count</c>.
/// </summary>
public static class SegmentBoundaryOffset
{
    public static int UnicodeOffset(string text, int utf16Index)
    {
        if (string.IsNullOrEmpty(text)) return 0;
        var index = Math.Clamp(utf16Index, 0, text.Length);
        if (index > 0 && index < text.Length && char.IsLowSurrogate(text[index]))
            index--;
        return text[..index].EnumerateRunes().Count();
    }

    public static int Utf16Index(string text, int unicodeOffset)
    {
        if (string.IsNullOrEmpty(text) || unicodeOffset <= 0) return 0;
        var remaining = unicodeOffset;
        var utf16 = 0;
        foreach (var rune in text.EnumerateRunes())
        {
            if (remaining <= 0) break;
            utf16 += rune.Utf16SequenceLength;
            remaining--;
        }
        return utf16;
    }

    public static (string Left, string Right) Split(string text, int unicodeOffset)
    {
        text ??= "";
        if (unicodeOffset <= 0) return ("", text);
        var utf16 = Utf16Index(text, unicodeOffset);
        if (utf16 >= text.Length) return (text, "");
        return (text[..utf16], text[utf16..]);
    }

    /// <summary>
    /// Matches Python <c>snap_cut_offset</c>: nearest candidate, ties go to the smaller offset.
    /// </summary>
    public static int NearestOffset(int offset, IReadOnlyList<int> candidates)
    {
        if (candidates is null || candidates.Count == 0) return offset;
        var best = candidates[0];
        var bestKey = (Math.Abs(best - offset), best);
        for (var i = 1; i < candidates.Count; i++)
        {
            var item = candidates[i];
            var key = (Math.Abs(item - offset), item);
            if (key.CompareTo(bestKey) < 0)
            {
                best = item;
                bestKey = key;
            }
        }
        return best;
    }

    public static bool CanSave(int previewCut, int originalCut, int totalChars, bool isSaving) =>
        !isSaving && previewCut != originalCut && previewCut > 0 && previewCut < totalChars;
}
